### ISSUE TITLE
Tooling, testing and CI enhancement proposal

### ISSUE THREAD

**Issue by silvanocerza** [supercollider/supercollider#6340]
Hello hello everyone! After the discussion that I had with @capital-G on #6319 regarding pre-commit hooks we had a really nice chat together about the project CI and other related stuff.

I decided I'd make a proposal on how to enhance it so here it is. This is really high level, if there needs be I can outline in more details what I want and can do.

## Tooling & Linting

The project is using a pretty old version of `clang-format`, it should be updated.
@capital-G told there's been some friction on this cause of huge formatting changes that caused some merge conflicts.

I think we can work around that by making the change gradual and having formatting checks only on files that are actually modified in future PRs.

I'd advocate to remove the version requirement set in `tools/clang-format.py` and deprecate it right after, as it makes it hard to use the actual `clang-format`. 

Change CI and pre-commit hooks so they check only files that have actually been modified in the branch or PR that is.

This will enable us to update the rules in `.clang-format` to support the latest ones and gradually enhance the project formatting without causing to much conflicts.

## Testing

I see that testing is done using [qpm](https://github.com/supercollider/qpm), a pretty outdated internal library that still runs on Python 2.7.

We should find a solution to that, either find a new way to run tests or rewrite the test runner for Python 3.

@capital-G also mentioned to me that some tests are flaky, we would need to find a way to mark those tests and rerun them some time before considered them completely failed.
## CI

The CI needs some improvement. 

Main problem I see right now is that there's a single `actions.yml` workflow that is quite convoluted and complex. 

First thing I would do is split that into multiple reusable workflows that are called by the existing one. This would already give us some breathing room.

This way we can rework the workflows so that we don't need to wait for all the builds to end to run tests. I'd probably go and split the building and testing process into 3 workflows, one for each platform. I'll figure it out as I go probably. In any case I won't make any change in the logic for the time being.

Splitting the `actions.yml` workflow we can also enhance the release process obviously.

Having said that I would like to hear any concerns from the usual contributors and go from there.
If there are no issues with the approach I propose I can take this into my hand and work on it while documenting it obviously.

**Comment by silvanocerza** [comment]
> v14.0 is the latest stable version distributed via debian and we recently updated from 8 to 14 - I think we can stay on 14 for a few years? The main goal of the formatter should be that it does not produce problems.
> 
> Because formatting rules differ between clang-format versions we need to stick to one. I think by switching to pre-commit we solved this in a nice manner.

I would upgrade in any case, using the [precompiled Python wheels of `clang-format`](https://pypi.org/project/clang-format) should make it trivial to move to latest. They also support Linux, MacOS and Windows so that shouldn't be a blocker.

Also with latest Debian you can install `clang-format` up to 16. These are the available `clang-format` packages I found on [the list of official stable packages](https://packages.debian.org/stable/allpackages?format=txt.gz). So if we need to stick to least common denominator we can at least go up to 16. I would go with the previous solution though.

```
clang-format (1:14.0-55.7~deb12u1) Tool to format C/C++/Obj-C code
clang-format-13 (1:13.0.1-11+b2) Tool to format C/C++/Obj-C code
clang-format-14 (1:14.0.6-12) Tool to format C/C++/Obj-C code
clang-format-15 (1:15.0.6-4+b1 [amd64, arm64, armel, armhf, i386, mipsel, ppc64el, s390x], 1:15.0.6-4 [mips64el]) Tool to format C/C++/Obj-C code
clang-format-16 (1:16.0.6-15~deb12u1) Tool to format C/C++/Obj-C code
```

> I'd support in deprecating the formatting script in favor of a pre-commit native workflow, as it creates benefits (faster runs) and offers us to delete code which otherwise needs to be maintained. Though I must admit that if the advanced features of the formatting script may be necessary for e.g. a release?

I still need to fully investigate the use of the script, though if some parts are necessary for the release or other important stuff they can be kept and moved in a separate ad hoc script in my opinion. The rest should be scrapped.

> Actually I think we are not using qpm anymore? I've made [a Python3 PR](https://github.com/supercollider/qpm/pull/14) for it but the project was archived soon afterwards. I don't know if it is necessary for windows test runs though.

My bad, you're right. I was looking at `main` branch instead of `develop`. 🤦 

> The idea is that, because, some tests are "unreliable" because of the container environment of GHA, we simply repeat a run on the failed tests. A test which is "wrong" will always be wrong, but tests which are only occasionally failing can be given a 2nd chance. I think that can give us some more reliable testing for CI.
> 
> See also #5360

That issue is interesting, seems like it could also be that tests are not isolated from each other. 🤔 
I'll look into that and see if isolation is enough to solve the problem. Otherwise I'll find a way to mark certain tests as flaky and rerun them a couple of times if they fail before considering them failed for real.

> Looking forward to your contributions!

Thanks! Should I create tracking issues or should I go ahead and start creating PRs?

In any case I prefer if we merge #6319 before that as the first thing that I'd like to do is split the `actions.yml` workflow and it would cause a conflict there.

**Comment by smoge** [comment]
> I think we can work around that by making the change gradual and having formatting checks only on files that are actually modified in future PRs.

I think that is much better than a "wall" hiding much of the git history. I would suggest that only files that are modified and that the "functional" commit with the code changes be separated from another commit for pretty printing in that file. Some projects follow this last detail and I think it makes sense when we need to study the history of some files.

**Comment by silvanocerza** [comment]
> I would suggest that only files that are modified and that the "functional" commit with the code changes be separated from another commit for pretty printing in that file if other unrelated lines are changed.

That's the idea. Separating formatting from actual code changes though is really hard to check with tooling, that should be on the contributor to make that separation in my opinion.

> That is much better than a "wall" hiding much of the git history.

That's debatable in my opinion, the history is still there and there are several ways to hide certain commits when calling different commands. It could actually be even easier to reformat everything in a single commit than gradually. 

As an example `git blame` can use [the flag `--ignore-rev`](https://git-scm.com/docs/git-blame#Documentation/git-blame.txt---ignore-revltrevgt) to ignore a single commit. 

`git log` has different ways instead. There's [`--exclude`](https://git-scm.com/docs/git-log#Documentation/git-log.txt---excludeltglob-patterngt), thought that would require adding some ref, maybe a tag, on the formatting commit. [`--invert-grep`](https://git-scm.com/docs/git-log#Documentation/git-log.txt---invert-grep) is another option, that checks the commit message though.

You'd have to know that prior obviously, but it shouldn't cause any issue when investigating bugs. You can easily skip it with `git bisect`, assuming formatting didn't break anything. 😅 

The only downside I think would be in the really short period as it would cause conflicts in all open PRs. But that can be handled, maybe with some difficulties but still worth it.

**Comment by smoge** [comment]
> As an example `git blame` can use [the flag `--ignore-rev`](https://git-scm.com/docs/git-blame#Documentation/git-blame.txt---ignore-revltrevgt) to ignore a single commit.

Yes, I didn't say it was not possible. But those things are obstacles to something that otherwise would be immediate.

**Comment by capital-G** [comment]
> I would suggest that only files that are modified and that the "functional" commit with the code changes be separated from another commit for pretty printing in that file if other unrelated lines are changed.

I'd suggest not to do that, quoting from https://github.com/readme/guides/code-review

> The last approach is an "on touch" incremental adoption.  In this mode, the quality tooling is left in an unenforced mode but enabled locally.  When a developer touches a file the tooling requires changes to be made to bring the code in line with standards.  his is often the easiest to adopt because it requires little to no infrastructure investment, but it leads to an undesirable result: the actual changes are obfuscated by unrelated formatter / linter changes.  For that reason, this approach isn't recommended.

Additionally, it makes it really hard (maybe even impossible?) to verify the formatting via CI, which is definitely something we don't want after we decided to introduce a formatter.

In general I am not sure why we should just bump clang-format for no reason, this will only generate unnecessary noise. The recent bump to clang-format-14 was necessary because the prior version clang-format-8 was not easily available on macOS anymore. We should bump it only if there is a reason to do so.

**Comment by smoge** [comment]
> I'd suggest not to do that, quoting from https://github.com/readme/guides/code-review

Who is this person? There are no arguments and no context in this article; he repeats the same thing as it would make it "true".)))))))

Also, this is a 20-year project.

**Comment by JordanHendersonMusic** [comment]
> Who is this person?

They made pre-commit.

**Comment by smoge** [comment]
> > Who is this person?
> 
> They made pre-commit.

Of course!)))

**Comment by capital-G** [comment]
> Of course! But there are no arguments there.

I think I quoted the argument and the whole article is pretty concise and clear on the topic. Using incremental updates taint regions in the code/file that are not affected by the actual commit/change to occur. This is arguable not good because this suggest that "commit A broke B" b/c the line changes are linked via the commit, but it is only linked due to formatting changes. With a big "blob" of formatting you can adjust git blame etc. to ignore said commit (see above), and also the commit message tells you "this was only changed b/c of formatting".

> For example, the guideline "stick to the same coding style already used in the file you are editing" has a rationale. Reformatting several times a 20-year codebase would need one as well.

The problem is that there is no code style beyond the C++ codebase and many comments on reviews are about sclang styling issues which should not be the case.
The community needs to decide what kind of formatting rules should be applied, but once we go the path of formatting, we should not leave it and verify the state via CI - this made the clang-format 14 update necessary.

**Comment by JordanHendersonMusic** [comment]
> The problem is that there is no code style beyond the C++ codebase and many comments on reviews are about sclang styling issues which should not be the case.

Once I'm through working on kwargs, I'm going to have a go at writing a formatter for sclang... don't know how well it will go. I'd be in favour of one big formatting commit.

**Comment by smoge** [comment]
> Once I'm through working on kwargs, I'm going to have a go at writing a formatter for sclang... don't know how well it will go. I'd be in favour of one big formatting commit.

What happened in the last years? It was never a "thing" before.

> 
> Linux kernel coding style
> 
> This is a short document describing the preferred coding style for the linux kernel. Coding style is very personal, and I won’t force my views on anybody, but this is what goes for anything that I have to be able to maintain, and I’d prefer it for most other things too. Please at least consider the points made here.

**Comment by capital-G** [comment]
> {JOKE}
> 
> > We like to run the formatter constantly and impose a strict style, but we forgot to write the tests and document the codebase. There are almost no comments. Fashion precedes correctness.

Your mixing up two things here - documentation and code styling. Everyone is always invited to add and improve the documentation and add comments to existing code, but we are talking about formatting/styling problems here.

sclang also has indeed an official code styling guide (see https://github.com/supercollider/supercollider/wiki/Style-Guidelines%3A-SuperCollider), but it is not enforced, or if it is enforced, it is enforced manually which is a bit annoying I think.

I personally don't have any strong feelings if we should impose it on the whole class lib, but having a good formatter would be really nice beyond that, e.g. it could be used within the IDE.

> Once I'm through working on kwargs, I'm going to have a go at writing a formatter for sclang... don't know how well it will go. I'd be in favour of one big formatting commit.

Maybe we can join efforts here?
If we keep the formatter within C++ territory it could replace the existing formatter in the IDE as well.
We then could additionally distribute the formatter as a python wheel (like the clang-format wheel), so the formatter does not have to be compiled manually (which would be hard for people who only work on class lib) but can be simply installed via pip.
I also thought about writing a Python binding for the SCDoc and sclang parser for Baryon.

**Comment by smoge** [comment]
It's a good thing to have a sclang formatter. However, I question strictly imposing it everywhere all the time, as it would be very beneficial.

**Comment by JordanHendersonMusic** [comment]
> it is enforced manually which is a bit annoying I think.

The majority of the pr review comments I get are about this. It is important, but definitely annoying, and takes from the people time that should have been spent looking at the implementation. For someone not intimately familiar with the styling guide (and who should be, its rather long) doing this automatically would be great and save their time and reviewers. 

> it could be used within the IDE.

I press `ctl+A, tab` (and `:w`) every few seconds in the supercollider IDE to format the code. When working with the class library, once I'm done, I have to sift through all the changes making sure the indentor has changed some old code. This is also annoying. For that reason, I'd be for a class library reformat and making sure it follows the guidelines particularly if the commit can be ignored from the blame.

> Maybe we can join efforts here?

Sure! I don't have time right now to do this, but I'll perhaps in the autumn. I was thinking about using sparkle and antlr which I think has support for formatting. 

I would like to make sure the formatter has two versions though, a strict mode for PRs and CI, and a series of options the user can select. Making the tool useful to those who have several decades of code that potentially predates the style guidelines is important.

**Comment by smoge** [comment]
> I also thought about writing a Python binding for the SCDoc and sclang parser for Baryon.

The best practice would be to use the language's parser to implement the formater (not possible right now, the correct order would be rewrite the language parser maybe). Extensive testing to check idempotence and ensure no parsing errors.

Also, a good implementation should allow the user to exercise some control while still guaranteeing that the formatted code is stylistically consistent.

**Comment by smoge** [comment]
> I press `ctl+A, tab` (and `:w`) every few seconds in the supercollider IDE to format the code. When working with the class library, once I'm done, I have to sift through all the changes making sure the indentor has changed some old code. This is also annoying. For that reason, I'd be for a class library reformat and making sure it follows the guidelines particularly if the commit can be ignored from the blame.

I think I like emacs for scaling code precisely because it does a much better job regarding formatting.

Also, maybe you guys think I hate formaters, but I happen to use the most conservative formater for Haskell (ormolu) in all my projects. 

My comments here are regarding SC  20-year codebase and context.

**Comment by smoge** [comment]
> > Once I'm through working on kwargs, I'm going to have a go at writing a formatter for sclang... don't know how well it will go. I'd be in favour of one big formatting commit.
> 
> Maybe we can join efforts here?

I was thinking of writing a parser (and a quasiquoter) in Haskell just for SynthDef as written in sclang. Maybe there is some overlap? It will be limited to SD, and will output the bytecode.

Maybe that's too far from what you guys will do.

**Comment by telephon** [comment]
> I was thinking of writing a parser (and a quasiquoter) in Haskell just for SynthDef as written in sclang.

Like this? https://hackage.haskell.org/package/hsc3-0.20/docs/Sound-Sc3-Server-Synthdef.html

**Comment by smoge** [comment]
> > I was thinking of writing a parser (and a quasiquoter) in Haskell just for SynthDef as written in sclang.
> 
> Like this? https://hackage.haskell.org/package/hsc3-0.20/docs/Sound-Sc3-Server-Synthdef.html

No, with the exact sclang syntax. The result would be that (I will use the hard part of the code from there, like the topological sorting ).

But my idea is to just write synthdefs as I write in sclang

**Comment by smoge** [comment]
Something like this:

```
[synthdef|
SynthDef(\help_notRand, { |out|
    Out.ar(out,
        SinOsc.ar(rrand(400, 800), 0, 0.2) * Line.kr(1, 0, 1, doneAction: Done.freeSelf)
    )
}).add|]
```

**Comment by smoge** [comment]
In the style of in-line C quasiquotation:

```
[C.block| int {
      // Read and sum 5 integers
      int i, sum = 0, tmp;
      for (i = 0; i < 5; i++) {
        scanf("%d", &tmp);
        sum += tmp;
      }
      return sum;
    } |]
```

**Comment by telephon** [comment]
Looks good (of course!). The question with embedded languages is always how the host language may (partly) re-appear in the embedded one, or how they are integrated with each other.

**Comment by smoge** [comment]
> Looks good (of course!). The question with embedded languages is always how the host language may (partly) re-appear in the embedded one, or how they are integrated with each other.

SynthDef is already a dialect of sclang. Multichannel expansion is something that sclang will bring. 

Faust operators (:> and <: etc) are also interesting.

**Comment by smoge** [comment]
The APL Bounded Loop operator is also very block diagram-like

```
{AllpassN.ar(_,0.050,[Rand(0,0.05),Rand(0,0.05)],1)}⍣4

instead of

4.do({ source = AllpassN.ar(source, 0.050, [Rand(0, 0.05), Rand(0, 0.05)], 1) })

```

**Comment by silvanocerza** [comment]
I see I stirred something with formatting. 😅 

Just to be clear that I'm just talking about formatting of C/C++ code in C/C++ files as of now. 
`sclang` files are not on my target as of now. Though I don't see any issue in formatting them if there's an actual standard.

> Also, this is a 20-year project; there is precious information in the history.

I don't understand this comment really. The history is not rewritten, where's the problem?

> Who is this person? There are no arguments and no context in this article; he repeats the same thing as if it would make it "true".)))))))

There's a really nice argument in there right at the end of the article. It maks code review faster cause you don't need to worry about any else other than the logic itself. Disregard the "fun" part.

> Adoption of code quality tooling can be arduous and disruptive, but it’s well worth the effort. I now rarely spend time in code reviews on style and conventions—that responsibility falls almost entirely to computers.  When I identify opportunities to make style improvements, instead of conversations about the more pedantic aspects of programming, I now think about how to fix the issue with tooling. It’s a shift that makes code reviews more effective — and a lot more fun.

**Comment by telephon** [comment]
Yes, sclang formatting is a complicated topic, because it is (also) about artistic style.

**Comment by JordanHendersonMusic** [comment]
@capital-G @silvanocerza 

Could the CI tests of SCPlugins be optionally ran when code relating to the server is touched?

**Comment by silvanocerza** [comment]
> Yes, sclang formatting is a complicated topic, because it is (also) about artistic style.

Makes totally sense and I see no issue with that. 

Introducing a formatter shouldn't force anyone to use in their own projects. Though I would force its use in test files used in this project. In my view those are supposed to be useful scripts to verify the correct behaviour of the project.

> Could the CI tests of SCPlugins be optionally ran when code relating to the server is touched?

I'm not an expert in the code base so I'm not sure what code is server related, but if it's clearly defined only in certain files that's certainly feasible. 

If necessary we could also make it so that adding a certain label to a PR triggers tests for SCPlugins. 

There are different approaches, this is an optimization that I would leave for later. Splitting `actions.yml` is my priority.

**Comment by telephon** [comment]
Yes, for test files, why not.

I wouldn't want to discuss it all here, but one argument for not enforcing it in the class library could be that you see the variety of valid syntax in the living source. But well.

A counter argument for enforcing it for tests and common sources is that it makes it harder to contribute, because you have to set up the pre-commit (I think?).

**Comment by JordanHendersonMusic** [comment]
@telephon 
> Yes, for test files, why not. 

@silvanocerza 
> Though I would force its use in test files used in this project.

I think that having a formatted class library would be beneficial because it makes reading the code easier through consistency — this is code to read by both developers and sometimes users when they want/need to understand how something works. Also cutting down on review time.

Whereas tests, in some cases, might want to disregard the formatter, particularly when it changes semantics, e.g. `arg ` -> `| |`, and `x.if(...)` -> `if(x) {...} {...}`. So there might need to be some granularity in how/where the formatter is applied.

**Comment by smoge** [comment]
> Whereas tests, in some cases, might want to disregard the formatter, particularly when it changes semantics, e.g. `arg ` -> `| |`, and `x.if(...)` -> `if(x) {...} {...}`. So there might need to be some granularity in how/where the formatter is applied.
> 
> I'm suggesting the formatter should have a big list of options that the user can toggle if they find the strict style of the class library doesn't work for them.

Someone created a "FaustGPT"; it's a matter of time before someone else makes a "SuperColliderGPT". Since their source is exactly the documentation and things like that, all this conversation can, in the long term, have unpredictable consequences. 

https://www.yeschat.ai/gpts-2OToJMDtFX-FAUSTGPT

![image](https://github.com/supercollider/supercollider/assets/271230/69f95cae-d98b-47ec-9b25-85f822f5fd09)

**Comment by telephon** [comment]
@smoge this comment has nothing to do with the issue. Could you *please, please* make an effort to stay on topic?

**Comment by telephon** [comment]
OK, sorry, maybe I am wrong. It is a little off though, but not too much. But let's keep it a little focused, ok?

**Comment by smoge** [comment]
Ok, I just pointed out one more thing that will contribute to standardization long-term. SC is not Python, it never was.

**Comment by dyfer** [comment]
@silvanocerza @capital-G just a quick note on tests:
As it @capital-G pointed out, we are not using qpm anymore. I moved the tests to run directly from command line:  https://github.com/supercollider/supercollider/pull/6003
Test themselves were/are written in sclang anyway. The configuration was previously json, now it's just an SC array of dictionaries.

Testing system is far from ideal, just wanted to chime in on the last changes.
Also I was tracking some of the issues with failing tests: #5359, #5360, #5421. Might be useful for reference...

**Comment by JordanHendersonMusic** [comment]
Something has changed with the CI that doesn't look right. 

It seems a test is ran on pull request and on commit? But the pull request one doesn't update correctly or something I don't really know whats going on.

https://github.com/supercollider/supercollider/pull/6365 Here I noticed that the pr ones would fail, but the push ones wouldn't?

**Comment by silvanocerza** [comment]
@dyfer @capital-G @JordanHendersonMusic 

I have some doubts about the `deploy_s3` job. 

From my understanding that's supposed to create a test build from the main branch (`develop`) of the repo. 

Though given the way the job is defined that's not true and is also possible to release from any other branch pushed in this repo, forks are excluded. In fact this already happened, see [this workflow run](https://github.com/supercollider/supercollider/actions/runs/9611396170) as an example. Notice how the `deploy_s3` job has been ran.

The main question would be when should this be run? Only when PRs are merged in `develop`? Or also for PRs?

I can make it work both ways, though it won't be possible to upload PR builds from forks since they don't have access to secrets. 

Also I'm not sure why the Linux build is completely ignored for S3 upload.

**Comment by dyfer** [comment]
TL;DR We can certainly discuss the need for S3 deployment on non-protected branches. I think the main point for it is that it's publicly accessible, as opposed to the GH artifacts, which require logging in to GH.

-------

> I have some doubts about the `deploy_s3` job.

For reference, @scztt was the one who I believe originally set up S3 deployment, even before moving to GitHub Actions, i.e. at the times we used AppVeyor and Travis.

> From my understanding that's supposed to create a test build from the main branch (`develop`) of the repo.

IIRC correctly, the idea was to easily get builds from the main repo, from any branch. While GH provides artifacts, one needs to be logged to GH to get them (I know there are workarounds, but I believe this is still the user-facing functionality through the web interface). GH artifacts vs S3 also have different retention time (though I don't know what's the retention policy on S3, as I think that was configured by @scztt )

I _think_ one reason for this was to easier get builds for testing etc. This might not be as relevant anymore and we can certainly discuss the path forward, but I want us to have a good understanding of what the existing approach is. Also, I'm writing this from memory, so other devs please chime in. I want to make sure that we run any changes to S3 by @scztt.

> 
> Though given the way the job is defined that's not true and is also possible to release from any other branch pushed in this repo, forks are excluded. In fact this already happened, see [this workflow run](https://github.com/supercollider/supercollider/actions/runs/9611396170) as an example. Notice how the `deploy_s3` job has been ran.
> 
> The main question would be when should this be run? Only when PRs are merged in `develop`? Or also for PRs?

Again, the idea was that S3 deployment is run for push to any branch to the repo. So not necessarily on PRs - the upload does not happen on PRs made from forks, which I think is the majority of them.

> I can make it work both ways, though it won't be possible to upload PR builds from forks since they don't have access to secrets.

The original idea was to upload only from the main repo, and not the forks.

> 
> In my opinion we should run that only in `develop`, the PRs have ways to get the binaries for testing in any case using the uploaded artifacts.

The idea was that for testing the build doesn't have to be on `develop` and I think that's a good thing - if one wants to test something, it doesn't have to go on develop. Actually, maybe it shouldn't be on develop until it's tested :) Of course depending on what kind of testing we're talking about, how mature is the functionality etc etc.
 
> Also I'm not sure why the Linux build is completely ignored for S3 upload.

That's because Linux builds are not distributable, fully functional "portable" builds. They are useful for testing (maybe?), but can't easily run on any distro, as they require dependencies specific to the GH build environment (i.e. Ubuntu). If we made e.g. an AppImage, it would make sense to also upload that.

**Comment by silvanocerza** [comment]
Ok, that's interesting. 

If the purpose is to distribute only binaries for testing purpose I think we can rely on workflow artifacts only. 

I would say the fact that you must login to download them is not a problem, I would expect someone creating a PR to have no issues from there. Their retention is 90 days by default and it can be lowered if necessary. I don't see the retention time as problem if they're meant to be used for testing only, 3 months should be enough. 😅 

At that point workflow artifacts would become a one stop shop for all test builds. I think contributors opening PRs from forks are already relying on those only given that the binaries can't be uploaded from those workflow runs. We can even make them more prominent in a PR by creating a comment that lists them out. 

Given all this I would drop the job completely. This should also cut some costs I guess, S3 doesn't come for free. :)

I'll wait for more info from other people in any case.

**Comment by dyfer** [comment]
> I would say the fact that you must login to download them is not a problem, I would expect someone creating a PR to have no issues from that

The issue comes up when the author seeks testing from the larger community, e.g. on the forum.

Also, S3 is used to host the latest `develop` build; these are linked from our website and IMO should keep being accessible without GH login.

https://nightly.link/ provides a way to download artifacts without logging in, but it's not as elegant (there's no permanent link to the artifact like we have with our download link) and since it works - AFAIU - against GH design, I'm not sure if we should depend on such functionality.

I personally don't mind moving away from S3, but I think we should maintain downloads without GH login, for sure for the develop builds.

I hope @scztt will chime in on this.
